pkg/reconcilers/deployment_mutator.go

```diff
@@ -88,17 +93,63 @@ func DeploymentImageMutator(desired, existing *k8sapps.Deployment) bool {
 	return update
 }
 
-func DeploymentVolumesMutator(desired, existing *k8sapps.Deployment) bool {
+func DeploymentImagePullPolicyMutator(desired, existing *k8sapps.Deployment) bool {
 	update := false
 
-	if !reflect.DeepEqual(existing.Spec.Template.Spec.Volumes, desired.Spec.Template.Spec.Volumes) {
-		existing.Spec.Template.Spec.Volumes = desired.Spec.Template.Spec.Volumes
+	if existing.Spec.Template.Spec.Containers[0].ImagePullPolicy != desired.Spec.Template.Spec.Containers[0].ImagePullPolicy {
+		existing.Spec.Template.Spec.Containers[0].ImagePullPolicy = desired.Spec.Template.Spec.Containers[0].ImagePullPolicy
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentContainerArgsMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	existingArgs := existing.Spec.Template.Spec.Containers[0].DeepCopy().Args
+	desiredArgs := desired.Spec.Template.Spec.Containers[0].DeepCopy().Args
+
+	existingArgsSortable := sort.StringSlice(existingArgs)
+	existingArgsSortable.Sort()
+	desiredArgsSortable := sort.StringSlice(desiredArgs)
+	desiredArgsSortable.Sort()
+
+	if strings.Join(existingArgsSortable, " ") != strings.Join(desiredArgsSortable, " ") {
+		existing.Spec.Template.Spec.Containers[0].Args = desired.Spec.Template.Spec.Containers[0].Args
 		update = true
 	}
 
 	return update
 }
 
+func DeploymentVolumesMutator(desired, existing *k8sapps.Deployment) bool {
+	existingVolumes := existing.Spec.Template.Spec.DeepCopy().Volumes
+	desiredVolumes := desired.Spec.Template.Spec.DeepCopy().Volumes
+
+	if len(existingVolumes) != len(desiredVolumes) {
+		existing.Spec.Template.Spec.Volumes = desired.Spec.Template.Spec.Volumes
+		return true
+	}
+
+	sort.Slice(existingVolumes, func(i, j int) bool {
+		return existingVolumes[i].Name < existingVolumes[j].Name
+	})
+
+	sort.Slice(desiredVolumes, func(i, j int) bool {
+		return desiredVolumes[i].Name < desiredVolumes[j].Name
+	})
+
+	for i, desiredVolume := range desiredVolumes {
+		if existingVolumes[i].Name != desiredVolume.Name { // comparing only the names has limitation, but more reliable than using reflect.DeepEqual or comparing the marshalled version of the resources
```

Comment:
Hmm not sure how I feel about this. Is there a better way of documenting this limitation or are we better off also iterating through the sources to update when sources change aswell ?